openssl/src/ssl/mod.rs

```diff
@@ -1471,6 +1505,48 @@ impl SslContextBuilder {
         }
     }
 
+    /// Adds a custom extension for a TLS/DTLS client or server for all supported protocol versions.
+    ///
+    /// This corresponds to [`SSL_CTX_add_custom_ext`].
+    ///
+    /// [`SSL_CTX_add_custom_ext`]: https://www.openssl.org/docs/manmaster/man3/SSL_CTX_add_custom_ext.html
+    #[cfg(all(feature = "v111", ossl111))]
+    pub fn add_custom_ext<AddFn, ParseFn>(&mut self, ext_type: u16, context: ExtensionContext, add_cb: AddFn, parse_cb: ParseFn)
+                                          -> Result<(), ErrorStack>
+        where AddFn: Fn(&mut SslRef, ExtensionContext, Option<(usize, &mut X509Ref)>)
+                        -> Result<Option<Cow<'static, [u8]>>, SslAlert> + 'static + Sync + Send,
+              ParseFn: Fn(&mut SslRef, ExtensionContext, &[u8], Option<(usize, &mut X509Ref)>)
```

Comment:
`&mut X509Ref` doesn't really give you anything beyond what a `&X509Ref` would since they're reference counted. Maybe it should be a shared reference?

The `Option<(usize, X509)>` is a bit weird but I guess makes sense.

Reply:
For some reason OpenSSL uses a pointer-to-mutable for these, despite using pointer-to-const elsewhere. Is it possible that `X509Ref` might someday grow `&mut self` methods?

Reply:
It's a bit awkward - we can't add &mut self methods to X509Ref directly since there can be multiple handles to the same X509 which would violate mutable aliasing guarantees. You could in theory have something like the `Arc::make_mut` or `Arc:get_mut` methods but then we'd need another type for the uniquely owned version. So far we just make builders for these kinds of types. `SslContext` and `X509StoreContext` come to mind as examples.

Reply:
Actually yeah, I think we may *have* to use `&X509Ref` here to avoid hitting Rust-side UB due to those aliasing concerns.